### ISSUE TITLE
Rule proposal: no-this-in-object-expression

### ISSUE THREAD

**Issue by axetroy** [sindresorhus/eslint-plugin-unicorn#2431]
### Description

It is unsafe to use this in an object's properties

It may contain potential bugs

```js
const obj = {
  name: 'foo',
  foo() {
    return this.name
  }
}

obj.foo() // 'foo'

const foo = obj.foo;

foo() // ''
```

### Fail

```js
const obj = {
  name: 'foo',
  foo() {
    return this.name
  }
}
```

### Pass

```js
const obj = {
  name: 'foo',
  foo() {
    return obj.name
  }
}
```

### Proposed rule name

no-this-in-object-expression

**Comment by fisker** [comment]
I'm not sure about this, it common pattern, and most time, it mean to used as method call, the problem is not how the object defined, but how it's used.

Wish we have [bind operator](https://github.com/tc39/proposal-bind-operator).

**Comment by fregante** [comment]
I don’t think it’s any less safe than using it in classes. A class method `obj.foo` is indistinguishable from an object property `obj.foo`, so in both cases one has to be aware of `this`. 

There’s some merit to avoiding `this` in objects, I think, but then there’s also situations where people might want to use `this` and explicitly bind the function, like `obj.getName.call(user)`

I can’t say I used any of these patterns in a very long time though.

**Comment by fisker** [comment]
How about `no-method-destructuring`?

Forbid destructure a method without `.bind()`

```js
// Bad
const foo = obj.foo;

// Good
const foo = obj.foo.bind(obj)

// Good
const foo = (...args) => obj.foo(...args)
```

**Comment by sindresorhus** [comment]
I don't think we should add this rule.

The unsafe part is not defining an object literal method with `this`. That's a normal JavaScript pattern, and `this` is intentionaly determined by how the function is called, not where it was defined. The actual bug is extracting or passing the method around unbound, for example `const foo = obj.foo` or `const {foo} = obj`.

We already have `unicorn/no-this-outside-of-class` for users who want the stricter style of only using `this` in classes. For the use site problem, `@typescript-eslint/unbound-method` is the right existing rule, because it can use type information to know whether a referenced property is actually a method.
